blaze-core/src/main/scala/org/http4s/blaze/websocket/Http4sWSStage.scala

```diff
@@ -0,0 +1,113 @@
+package org.http4s
+package blaze
+package websocket
+
+import fs2.async.mutable.Signal
+import org.http4s.websocket.WebsocketBits._
+
+import scala.util.{Failure, Success}
+import org.http4s.blaze.pipeline.stages.SerializingStage
+import org.http4s.blaze.util.Execution.{directec, trampoline}
+import org.http4s.{websocket => ws4s}
+
+import fs2.async
+import fs2._
+
+import pipeline.{TrunkBuilder, LeafBuilder, Command, TailStage}
+
+class Http4sWSStage(ws: ws4s.Websocket)(implicit val strategy: Strategy) extends TailStage[WebSocketFrame] {
+  def name: String = "Http4s WebSocket Stage"
+
+  // TODO Remove
+  def log[A](prefix: String): Pipe[Task, A, A] = _.evalMap{a => Task.delay {println(s"$prefix> $a"); a} }
+
+  private val deadSignal: Task[Signal[Task, Boolean]] = async.signalOf[Task, Boolean](false)
+
+  //////////////////////// Source and Sink generators ////////////////////////
+
+  def snk: Sink[Task, WebSocketFrame] = _.evalMap { frame =>
+    Task.async[Unit] { cb =>
+      channelWrite(frame).onComplete {
+        case Success(res)             => cb(Right(res))
+        case Failure(t @ Command.EOF) => cb(Left(t))
+        case Failure(t)               => cb(Left(t))
+      }(directec)
+    }
+  }
+
+  def inputstream: Stream[Task, WebSocketFrame] = {
+    val t = Task.async[WebSocketFrame] { cb =>
+      def go(): Unit = channelRead().onComplete {
+        case Success(ws)         => ws match {
+            case Close(_)    =>
+              for {
+                t <- deadSignal.map(_.set(true))
+              } yield {
+                // RFC It used to send the disconnect from here but we can use signal too
+                t.unsafeRun()
+                //sendOutboundCommand(Command.Disconnect)
+                cb(Left(Command.EOF))
+              }
+
+            // TODO: do we expect ping frames here?
+            case Ping(d)     =>  channelWrite(Pong(d)).onComplete {
+              case Success(_)           => go()
+              case Failure(Command.EOF) => cb(Left(Command.EOF))
+              case Failure(t)           => cb(Left(t))
+            }(trampoline)
+
+            case Pong(_)     => go()
+            case f           => cb(Right(f))
+          }
+
+        case Failure(Command.EOF) => cb(Left(Command.EOF))
+        case Failure(e)           => cb(Left(e))
+      }(trampoline)
+
+      go()
+    }
+    Stream.repeatEval(t) // RFC On the original code _.asHalt was called. Is it needed anymore?
```

Comment:
I don't think there's an `asHalt` equivalent, since fs2 doesn't use exceptions for control flow.